src/Core/Services/MultipleMutationInputValidator.cs

```diff
@@ -0,0 +1,575 @@
+            foreach (string columnDerivedFromSourceEntity in columnsDerivedFromSourceEntity)
+            {
+                derivableColumnsFromRequestBody.TryAdd(columnDerivedFromSourceEntity, parentEntityName);
+            }
+
+            // For the relationships with the parent entity, where the current entity is a referenced entity,
+            // we need to make sure that we have non-null values for all the referenced columns - since the values for all those
+            // columns will be used for insertion in the parent entity. We can get the referenced column value:
+            // Case 1. Via a scalar value provided in the input for the current entity
+            // Case 2. Via another relationship where a column referenced by the parent of the current entity is a
+            // referencing column for a target entity related with current entity.
+
+            // Case 1: Remove from columnsToBeDerivedFromEntity, the columns which are autogenerated or
+            // have been provided a non-null value in the input for the current entity.
+
+            // As we iterate through columns which can derived from the current entity, we keep removing them from the
+            // columnsToDerive because we know the column value can be derived.
+            // At the end, if there are still columns yet to be derived, i.e.columnsToDerive.count > 0,
+            // we throw an exception.
+            foreach (string columnToBeDerivedFromEntity in columnsToDerive)
+            {
+                if (sourceDefinition.Columns[columnToBeDerivedFromEntity].IsAutoGenerated)
+                {
+                    // The value for an autogenerated column is derivable.
+                    columnsToDerive.Remove(columnToBeDerivedFromEntity);
+                }
+                else if (backingColumnData.TryGetValue(columnToBeDerivedFromEntity, out IValueNode? value))
+                {
+                    metadataProvider.TryGetExposedColumnName(entityName, columnToBeDerivedFromEntity, out string? exposedColumnName);
+                    if (value is null)
+                    {
+                        throw new DataApiBuilderException(
+                            message: $"Value cannot be null for referenced field: {exposedColumnName} for entity: {entityName} at level: {nestingLevel}.",
+                            statusCode: HttpStatusCode.BadRequest,
+                            subStatusCode: DataApiBuilderException.SubStatusCodes.BadRequest);
+                    }
+
+                    columnsToDerive.Remove(columnToBeDerivedFromEntity);
+                }
+            }
+
+            // Dictionary storing the mapping from relationship name to the set of referencing fields for all
+            // the relationships for which the current entity is the referenced entity.
+            Dictionary<string, HashSet<string>> columnsDerivedForRelationships = new();
```

Comment:
```suggestion
            Dictionary<string, HashSet<string>> fieldsToBeSuppliedForReferencingEntities = new();
```
or something along these lines.. 

Suggestion to remove `derived` from this variable name as `derived` has an undertone of fetching PKs from the creation of other entities but the intent of the variable here is to hold PKs to supply to other entities